File: webflux-dialogue/src/main/java/com/study/webflux/rag/application/memory/service/MemoryExtractionService.java
package com.study.webflux.rag.application.memory.service;

import java.util.List;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import org.springframework.stereotype.Service;

import com.study.webflux.rag.domain.dialogue.model.ConversationTurn;
import com.study.webflux.rag.domain.dialogue.model.PersonaId;
import com.study.webflux.rag.domain.dialogue.model.UserId;
import com.study.webflux.rag.domain.dialogue.port.ConversationRepository;
import com.study.webflux.rag.domain.memory.model.ExtractedMemory;
import com.study.webflux.rag.domain.memory.model.Memory;
import com.study.webflux.rag.domain.memory.model.MemoryExtractionContext;
import com.study.webflux.rag.domain.memory.port.ConversationCounterPort;
import com.study.webflux.rag.domain.memory.port.EmbeddingPort;
import com.study.webflux.rag.domain.memory.port.MemoryExtractionPort;
import com.study.webflux.rag.domain.memory.port.VectorMemoryPort;
import com.study.webflux.rag.infrastructure.monitoring.config.MemoryExtractionMetricsConfiguration;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * 대화 이력과 기존 메모리를 바탕으로 신규 메모리를 추출/저장합니다.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class MemoryExtractionService {

	private final ConversationRepository conversationRepository;
	private final ConversationCounterPort counterPort;
	private final MemoryExtractionPort extractionPort;
	private final EmbeddingPort embeddingPort;
	private final VectorMemoryPort vectorMemoryPort;
	private final MemoryRetrievalService retrievalService;
	private final MemoryExtractionMetricsConfiguration extractionMetrics;
	private final int conversationThreshold;

	/**
	 * 누적 대화 수가 임계값 배수인 경우에만 메모리 추출 파이프라인을 실행합니다.
	 */
	public Mono<Void> checkAndExtract(PersonaId personaId, UserId userId) {
		return counterPort.get(personaId, userId)
			.filter(this::isExtractionTurn)
			.flatMap(count -> {
				log.info("메모리 추출 트리거: persona={}, 대화 횟수={}", personaId.value(), count);
				extractionMetrics.recordExtractionTriggered();
				return performExtraction(personaId, userId);
			})
			.then();
	}

	/**
	 * 하위 호환성을 위한 기존 메서드
	 */
	public Mono<Void> checkAndExtract(UserId userId) {
		return checkAndExtract(PersonaId.defaultPersona(), userId);
	}

	/**
	 * 최근 대화를 기준으로 메모리 추출 컨텍스트를 구성하고 추출 결과를 저장합니다.
	 */
	private Mono<Void> performExtraction(PersonaId personaId, UserId userId) {
		return loadRecentConversations(personaId, userId)
			.flatMap(conversations -> buildExtractionContext(personaId, userId, conversations))
			.flatMapMany(extractionPort::extractMemories)
			.collectList()
			.doOnNext(extractedList -> {
				extractionMetrics.recordExtractionSuccess(extractedList.size());

				java.util.Map<String, Long> typeCounts = extractedList.stream()
					.collect(java.util.stream.Collectors.groupingBy(
						extracted -> extracted.type().name(),
						java.util.stream.Collectors.counting()));

				typeCounts.forEach((type, count) -> extractionMetrics
					.recordExtractedMemoryType(type, count.intValue()));

				extractedList.forEach(extracted -> extractionMetrics
					.recordExtractedImportance(extracted.importance()));
			})
			.doOnError(error -> {
				extractionMetrics.recordExtractionFailure();
				log.error("메모리 추출 실패", error);
			})
			.flatMapMany(Flux::fromIterable)
			.flatMap(this::saveExtractedMemory)
			.doOnNext(memory -> log.info("추출 및 저장된 메모리: type={}, importance={}, content={}",
				memory.type(),
				memory.importance(),
				memory.content()))
			.then();
	}

	/**
	 * 추출된 메모리를 임베딩한 뒤 벡터 저장소에 저장합니다.
	 */
	private Mono<Memory> saveExtractedMemory(ExtractedMemory extracted) {
		Memory memory = extracted.toMemory();

		return embeddingPort.embed(memory.content())
			.flatMap(embedding -> vectorMemoryPort.upsert(memory, embedding.vector()));
	}

	private boolean isExtractionTurn(long count) {
		return count > 0 && count % conversationThreshold == 0;
	}

	/**
	 * 임계 대화 수만큼 최근 대화 이력을 불러옵니다.
	 */
	private Mono<List<ConversationTurn>> loadRecentConversations(PersonaId personaId,
		UserId userId) {
		return conversationRepository.findRecent(personaId, userId, conversationThreshold)
			.collectList();
	}

	/**
	 * 대화 이력과 관련 메모리를 결합해 추출 컨텍스트를 구성합니다.
	 */
	private Mono<MemoryExtractionContext> buildExtractionContext(PersonaId personaId,
		UserId userId,
		List<ConversationTurn> conversations) {
		String combinedQuery = mergeQueries(conversations);
		return retrievalService.retrieveMemories(personaId, userId, combinedQuery, 10)
			.map(result -> MemoryExtractionContext
				.of(personaId, userId, conversations, result.allMemories()));
	}

	private String mergeQueries(List<ConversationTurn> conversations) {
		return conversations.stream().map(ConversationTurn::query).reduce((a, b) -> a + " " + b)
			.orElse("");
	}
}

File: app-miyou/src/main/kotlin/com/miyou/app/application/memory/service/MemoryExtractionService.kt
package com.miyou.app.application.memory.service

import com.miyou.app.application.monitoring.port.MemoryExtractionMetricsPort
import com.miyou.app.domain.dialogue.model.ConversationSessionId
import com.miyou.app.domain.dialogue.model.ConversationTurn
import com.miyou.app.domain.dialogue.port.ConversationRepository
import com.miyou.app.domain.memory.model.ExtractedMemory
import com.miyou.app.domain.memory.model.Memory
import com.miyou.app.domain.memory.model.MemoryExtractionContext
import com.miyou.app.domain.memory.port.ConversationCounterPort
import com.miyou.app.domain.memory.port.EmbeddingPort
import com.miyou.app.domain.memory.port.MemoryExtractionPort
import com.miyou.app.domain.memory.port.VectorMemoryPort
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import java.time.Instant

/**
 * 메모리 추출 서비스.
 *
 * 주기적으로 대화 이력에서 의미있는 정보(인물, 선호도, 사건 등)를 추출하여 벡터 저장소에 저장.
 * 사용자 선호도 학습 및 개인화 맥락 강화 목적.
 */
@Service
class MemoryExtractionService(
    private val conversationRepository: ConversationRepository,
    private val counterPort: ConversationCounterPort,
    private val extractionPort: MemoryExtractionPort,
    private val embeddingPort: EmbeddingPort,
    private val vectorMemoryPort: VectorMemoryPort,
    private val retrievalService: MemoryRetrievalService,
    private val extractionMetrics: MemoryExtractionMetricsPort,
    private val conversationThreshold: Int,
) {
    private val logger = KotlinLogging.logger {}

    /**
     * 주기 도달 시 메모리 추출 수행.
     * 임계값(conversationThreshold) 도달마다 트리거됨.
     *
     * @param sessionId 대화 세션 ID
     * @return 추출 완료 (실패해도 무시)
     */
    fun checkAndExtract(sessionId: ConversationSessionId): Mono<Void> =
        counterPort
            .get(sessionId)
            .filter(::isExtractionTurn)
            .flatMap { count ->
                logger.info { "메모리 추출 트리거 sessionId=${sessionId.value}, count=$count" }
                extractionMetrics.recordExtractionTriggered()
                performExtraction(sessionId)
            }.then()

    private fun performExtraction(sessionId: ConversationSessionId): Mono<Void> =
        loadRecentConversations(sessionId)
            .flatMap { conversations ->
                buildExtractionContext(sessionId, conversations)
            }.flatMap { context -> extractAndSave(sessionId, context) }

    private fun extractAndSave(
        sessionId: ConversationSessionId,
        context: MemoryExtractionContext,
    ): Mono<Void> =
        extractionPort
            .extractMemories(context)
            .collectList()
            .doOnNext { extractedList ->
                extractionMetrics.recordExtractionSuccess(extractedList.size)

                val typeCounts =
                    extractedList
                        .groupingBy(ExtractedMemory::type)
                        .eachCount()

                typeCounts.forEach { (type, count) ->
                    extractionMetrics.recordExtractedMemoryType(type.name, count)
                }

                extractedList.forEach { extracted ->
                    extractionMetrics.recordExtractedImportance(extracted.importance.toDouble())
                    logger.info {
                        "메모리 추출 근거 sessionId=${sessionId.value}, type=${extracted.type}, " +
                            "importance=${extracted.importance}, reasoning=${extracted.reasoning.replace("\n", "\\n")}"
                    }
                }
            }.doOnError { error ->
                extractionMetrics.recordExtractionFailure()
                logger.error(error) { "메모리 추출 실패" }
            }.flatMapMany(Flux<ExtractedMemory>::fromIterable)
            .flatMap { extracted -> saveExtractedMemory(extracted, context.existingMemories) }
            .doOnNext { memory ->
                logger.info {
                    "추출된 메모리 저장 완료 type=${memory.type}, importance=${memory.importance}, content=${memory.content}"
                }
            }.then()

    private fun saveExtractedMemory(
        extracted: ExtractedMemory,
        existingMemories: List<Memory>,
    ): Mono<Memory> {
        val memory = extracted.toMemory()

        return embeddingPort
            .embed(memory.content)
            .flatMap { embedding ->
                vectorMemoryPort.upsert(memory, embedding.vector)
            }.flatMap { saved -> archiveSupersededMemory(extracted, existingMemories).thenReturn(saved) }
    }

    /** 모순된 기존 메모리를 즉시 소프트 아카이브 - 큐레이터가 이미 쓰는 archive 경로를 그대로 재사용한다. */
    private fun archiveSupersededMemory(
        extracted: ExtractedMemory,
        existingMemories: List<Memory>,
    ): Mono<Void> {
        val targetId = extracted.supersedesMemoryId ?: return Mono.empty()
        val target = existingMemories.firstOrNull { it.id == targetId } ?: return Mono.empty()
        return vectorMemoryPort
            .applyDecayAndArchive(target.archive(Instant.now()))
            .doOnSuccess {
                logger.info { "모순 감지로 기존 메모리 아카이브 id=$targetId, 대체 내용=${extracted.content}" }
            }
    }

    private fun isExtractionTurn(count: Long): Boolean = count > 0 && count % conversationThreshold == 0L

    private fun loadRecentConversations(sessionId: ConversationSessionId): Mono<List<ConversationTurn>> =
        conversationRepository
            .findRecent(sessionId, conversationThreshold)
            .collectList()

    private fun buildExtractionContext(
        sessionId: ConversationSessionId,
        conversations: List<ConversationTurn>,
    ): Mono<MemoryExtractionContext> {
        val combinedQuery = mergeQueries(conversations)
        return retrievalService
            .retrieveMemories(sessionId, combinedQuery, 10)
            .map { result ->
                MemoryExtractionContext.of(sessionId, conversations, result.allMemories())
            }
    }

    private fun mergeQueries(conversations: List<ConversationTurn>): String =
        conversations.joinToString(" ") { it.query }
}
